File: src/bin/0054-spiral-matrix.py
# @leetup=info id=54 lang=python3 slug=spiral-matrix



# @leetup=code


class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        R = len(matrix)
        C = len(matrix[0])

        low_r = 0
        high_r = R

        low_c = 0
        high_c = C

        r = 0
        c = 0

        res = []
        d = ((0, 1), (1, 0), (0, -1), (-1, 0))
        direction = 0
        while low_r <= r < high_r and low_c <= c < high_c:
            res.append(matrix[r][c])
            new_r, new_c = r + d[direction][0], c + d[direction][1]
            if low_r <= new_r < high_r and low_c <= new_c < high_c:
                r, c = new_r, new_c
            else:
                if direction == 0:
                    low_r += 1
                elif direction == 1:
                    high_c -= 1
                elif direction == 2:
                    high_r -= 1
                else:
                    low_c += 1
                direction = (direction + 1) % 4
                r, c = r + d[direction][0], c + d[direction][1]
        return res


# @leetup=code

File: src/bin/0054-spiral-matrix.cpp
// @leetup=info id=54 lang=cpp slug=spiral-matrix

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  std::vector<int> spiralOrder(std::vector<std::vector<int>>& matrix) {
    std::vector<int> res;

    size_t low_r = 0;
    size_t high_r = matrix.size();

    size_t low_c = 0;
    size_t high_c = matrix[0].size();

    size_t r = 0;
    size_t c = 0;

    size_t direction = 0;
    std::vector<std::tuple<int, int>> d{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

    while (low_r <= r && r < high_r && low_c <= c && c < high_c) {
      res.push_back(matrix[r][c]);
      size_t nr = r + std::get<0>(d[direction]);
      size_t nc = c + std::get<1>(d[direction]);
      if (low_r <= nr && nr < high_r && low_c <= nc && nc < high_c) {
        r = nr;
        c = nc;
      } else {
        if (direction == 0) {
          low_r += 1;
        } else if (direction == 1) {
          high_c -= 1;
        } else if (direction == 2) {
          high_r -= 1;
        } else {
          low_c += 1;
        }
        direction = (direction + 1) % 4;
        r += std::get<0>(d[direction]);
        c += std::get<1>(d[direction]);
      }
    }
    return res;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
